File: go/fixture/go-mini/internal/services/cache.go
package services

import (
	"fmt"
	"sync"
	"time"

	"example.com/go-mini/internal/models"
)

// Cache is a cache of devices.
type Cache struct {
	mu    sync.Mutex
	items map[string]models.Device
	ttl   map[string]time.Time
	span  time.Duration
}

// NewCache creates a new Cache whose entries expire after span.
func NewCache(span time.Duration) *Cache {
	return &Cache{
		items: map[string]models.Device{},
		ttl:   map[string]time.Time{},
		span:  span,
	}
}

func (c *Cache) Put(d models.Device) {
	key := cacheKey(d.Tenant, d.ID)
	c.mu.Lock()
	c.items[key] = d
	c.mu.Unlock()
	c.ttl[key] = time.Now().Add(c.span)
}

// Get returns the cached device and refreshes its expiry. Entries that have
// expired are dropped on the way.
func (c *Cache) Get(key string) (models.Device, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	now := time.Now()
	for k, exp := range c.ttl {
		if exp.Before(now) {
			delete(c.items, k)
			delete(c.ttl, k)
		}
	}
	d, ok := c.items[key]
	if ok {
		c.ttl[key] = now.Add(c.span)
	}
	return d, ok
}

// All returns every cached device.
func (c *Cache) All() map[string]models.Device {
	return c.items
}

// Len returns the number of cached devices.
func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.items)
}

// Footprint estimates the memory the cache holds, in bytes.
func (c *Cache) Footprint() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	items := c.items
	size := len(items)
	for _, d := range items {
		size += len(d.ID) + len(d.Tenant) + len(d.Version)
	}
	size = size * 10 // rough per-entry overhead
	size += len(c.ttl) * 24
	return size
}

// String renders the cache for logs.
func (c *Cache) String() string {
	return describe(*c) //nolint:govet // TODO
}

func describe(c Cache) string { //nolint:govet // TODO
	return fmt.Sprintf("cache: %d items, %d expiring", len(c.items), len(c.ttl))
}

File: py/fixture/py-mini/internal/services/cache.py
"""A TTL cache of devices."""

import copy
import threading
import time

from internal.models.device import Device
from internal.services.device_service import _cache_key


class Cache:
    """Cache is a cache of devices."""

    def __init__(self, span: float) -> None:
        """Create a Cache whose entries expire after span seconds."""
        self.mu = threading.Lock()
        self.items: dict[str, Device] = {}
        self.ttl: dict[str, float] = {}
        self.span = span

    def put(self, d: Device) -> None:
        """Store the device under its tenant and id."""
        key = _cache_key(d.tenant, d.id)
        with self.mu:
            self.items[key] = d
        self.ttl[key] = time.monotonic() + self.span

    def get(self, key: str) -> Device | None:
        """Return the cached device and refresh its expiry.

        Entries that have expired are dropped on the way.
        """
        with self.mu:
            now = time.monotonic()
            for k, exp in list(self.ttl.items()):
                if exp < now:
                    self.items.pop(k, None)
                    del self.ttl[k]
            d = self.items.get(key)
            if d is not None:
                self.ttl[key] = now + self.span
            return d

    def all(self) -> dict[str, Device]:
        """Return every cached device."""
        return self.items

    def count(self) -> int:
        """Return the number of cached devices."""
        with self.mu:
            return len(self.items)

    def footprint(self) -> int:
        """Estimate the memory the cache holds, in bytes."""
        with self.mu:
            items = self.items
            size = len(items)
            for d in items.values():
                size += len(d.id) + len(d.tenant) + len(d.version)
            size = size * 10  # rough per-entry overhead
            size += len(self.ttl) * 24
            return size

    def __str__(self) -> str:
        return _describe(copy.copy(self))


def _describe(c: Cache) -> str:
    with c.mu:
        return f"cache: {len(c.items)} items, {len(c.ttl)} expiring"
